src/cunumeric/sort/sort.cu

```diff
@@ -0,0 +1,691 @@
+      } else {
+        thrust::transform(thrust::cuda::par.on(stream),
+                          thrust::make_counting_iterator<int64_t>(0),
+                          thrust::make_counting_iterator<int64_t>(volume),
+                          thrust::make_constant_iterator<int64_t>(sort_dim_size),
+                          indices_buffer.ptr(0),
+                          thrust::modulus<int64_t>());
+      }
+    }
+
+    // sort data
+    local_sort_inplace<CODE>(dense_input_copy.ptr(0),
+                             argsort ? indices_buffer.ptr(0) : nullptr,
+                             volume,
+                             sort_dim_size,
+                             stream);
+
+    // this is linked to the decision in sorting.py on when to use adn 'unbounded' output array.
+    if (output_array.dim() == -1) {
+      SortPiece<VAL> local_sorted;
+      local_sorted.values                       = dense_input_copy;
+      local_sorted.indices                      = indices_buffer;
+      local_sorted.size                         = volume;
+      SortPiece<VAL> local_sorted_repartitioned = is_index_space
+                                                    ? sample_sort_nccl(local_sorted,
+                                                                       my_rank,
+                                                                       domain.get_volume(),
+                                                                       argsort,
+                                                                       stream,
+                                                                       comms[0].get<ncclComm_t*>())
+                                                    : local_sorted;
+      if (argsort) {
+        output_array.return_data(local_sorted_repartitioned.indices,
+                                 local_sorted_repartitioned.size);
+      } else {
+        output_array.return_data(local_sorted_repartitioned.values,
+                                 local_sorted_repartitioned.size);
+      }
+    } else {
+      // copy back data (we assume output partition to be aliged to input!)
+      if (argsort) {
+        AccessorWO<int64_t, DIM> output = output_array.write_accessor<int64_t, DIM>(rect);
+        assert(output.accessor.is_dense_row_major(rect));
+        cudaMemcpyAsync(output.ptr(rect.lo),
+                        indices_buffer.ptr(0),
+                        sizeof(int64_t) * volume,
+                        cudaMemcpyDeviceToDevice,
+                        stream);
+      } else {
+        AccessorWO<VAL, DIM> output = output_array.write_accessor<VAL, DIM>(rect);
+        assert(output.accessor.is_dense_row_major(rect));
+        cudaMemcpyAsync(output.ptr(rect.lo),
+                        dense_input_copy.ptr(0),
+                        sizeof(VAL) * volume,
+                        cudaMemcpyDeviceToDevice,
+                        stream);
+      }
+    }
+    CHECK_CUDA(cudaStreamSynchronize(stream));
```

Comment:
note that the runtime will make sure all kernels are done when the task exits, so you don't need to synchronize the stream here.